Code/Framework/AzCore/AzCore/DOM/DomValue.h

```diff
@@ -0,0 +1,385 @@
+
+        bool operator==(const Value& rhs) const;
+        bool operator!=(const Value& rhs) const;
+
+        void Swap(Value& other) noexcept;
+
+        // Type info...
+        Type GetType() const;
+        bool IsNull() const;
+        bool IsFalse() const;
+        bool IsTrue() const;
+        bool IsBool() const;
+        bool IsNode() const;
+        bool IsObject() const;
+        bool IsArray() const;
+        bool IsOpaqueValue() const;
+        bool IsNumber() const;
+        bool IsInt() const;
+        bool IsUint() const;
+        bool IsDouble() const;
+        bool IsString() const;
+
+        // Object API (also used by Node)...
+        Value& SetObject();
+        size_t MemberCount() const;
+        size_t MemberCapacity() const;
+        bool ObjectEmpty() const;
+
+        Value& operator[](KeyType name);
+        const Value& operator[](KeyType name) const;
+        Value& operator[](AZStd::string_view name);
+        const Value& operator[](AZStd::string_view name) const;
+
+        Object::ConstIterator MemberBegin() const;
+        Object::ConstIterator MemberEnd() const;
+        Object::Iterator MemberBegin();
+        Object::Iterator MemberEnd();
+
+        Object::Iterator FindMutableMember(KeyType name);
+        Object::Iterator FindMutableMember(AZStd::string_view name);
+        Object::ConstIterator FindMember(KeyType name) const;
+        Object::ConstIterator FindMember(AZStd::string_view name) const;
+
+        Value& MemberReserve(size_t newCapacity);
+        bool HasMember(KeyType name) const;
+        bool HasMember(AZStd::string_view name) const;
+
+        Value& AddMember(KeyType name, const Value& value);
+        Value& AddMember(AZStd::string_view name, const Value& value);
+        Value& AddMember(KeyType name, Value&& value);
+        Value& AddMember(AZStd::string_view name, Value&& value);
+
+        void RemoveAllMembers();
+        void RemoveMember(KeyType name);
+        void RemoveMember(AZStd::string_view name);
+        Object::Iterator RemoveMember(Object::Iterator pos);
+        Object::Iterator EraseMember(Object::ConstIterator pos);
+        Object::Iterator EraseMember(Object::ConstIterator first, Object::ConstIterator last);
+        Object::Iterator EraseMember(KeyType name);
+        Object::Iterator EraseMember(AZStd::string_view name);
+
+        Object::ContainerType& GetMutableObject();
+        const Object::ContainerType& GetObject() const;
+
+        // Array API (also used by Node)...
+        Value& SetArray();
+
+        size_t Size() const;
+        size_t Capacity() const;
+        bool Empty() const;
+        void Clear();
+
+        Value& operator[](size_t index);
+        const Value& operator[](size_t index) const;
+
+        Value& MutableAt(size_t index);
+        const Value& At(size_t index) const;
+
+        Array::ConstIterator Begin() const;
+        Array::ConstIterator End() const;
+        Array::Iterator Begin();
+        Array::Iterator End();
```

Comment:
While it goes against the coding style guide, using lower cases for these function names would allow them to be used in ranged-for loops.

Reply:
I think because we have both `MemberBegin` and `Begin` it'd actually be a bit confusing to use one of these in a ranged-for, but some kind of helper that returns an iterable might be cool.

Reply:
In that case consider being more explicit and call these `Array*` such as `ArrayBegin` or in that vein. I know this somewhat mirrors the RapidJSON API, but even there calls have been confusing. For instance "Empty" and "Clear" have been incorrectly used on objects. Calling these `IsArrayEmpty` and `ClearArray` will help remove that confusion.

Reply:
Agreed. Operations that operate on a children are tagged with `Array` now, I think the slight added verbosity is outweighed by the disambiguation from object operations we gain